src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -105,6 +104,24 @@ public PackageURL(final String type, final String namespace, final String name,
         verifyTypeConstraints(this.type, this.namespace, this.name);
     }
 
+    /**
+     * Constructs a new PackageURL object.
+     *
+     * @param type the type of package (i.e. maven, npm, gem, etc)
+     * @param namespace the name prefix (i.e. group, owner, organization)
+     * @param name the name of the package
+     * @param version the version of the package
+     * @param qualifiers an array of key/value pair qualifiers
+     * @param subpath the subpath string
+     * @throws MalformedPackageURLException if parsing fails
+     * @since 1.6.0
+     */
+    public PackageURL(final String type, final String namespace, final String name, final String version,
+                      final Map<String, String> qualifiers, final String subpath)
+            throws MalformedPackageURLException {
+        this(type, namespace, name, version, (qualifiers == null) ? null : ((qualifiers instanceof TreeMap) ? (TreeMap<String, String>) qualifiers : new TreeMap<>(qualifiers)), subpath);
```

Comment:
```suggestion
        this(type, namespace, name, version, (qualifiers == null) ? null : new TreeMap<>(qualifiers), subpath);
```

This will prevent the caller from unintentionally modifying the state of `PackageURL` by modifying the map.

Should this one be the canonical constructor that the one above should call?

Reply:
My version works to avoid making two copies of the map if the original constructor is making a copy, but it's not. I fixed it in #169 to parse the qualifiers and therefore make a copy.

The behavior of the existing code inconsistent. Right now, it's making a copy if you use the `String` constructor, but not making a copy if you use this constructor. I could close this and move this code to #169 in which case the way I have it written would make more sense, I think.

Reply:
Fixed in [a27618b](https://github.com/package-url/packageurl-java/pull/155/commits/a27618b518414b700436030fc0a565d7c3c57cba). But, #169 changes the code to make the qualifier parsing happen on both constructors so that the key/value pairs always get normalized.

Reply:
Probably it would make sense to merge this after #169.